File: src/autorace_perception/src/sensor_camera/per_camera.py
#!/usr/bin/env python3
#-*- coding:utf-8 -*- 
import sys
import os

current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.abspath(os.path.join(current_dir, '../..'))  # drive_controller 상위 폴더
if parent_dir not in sys.path:
    sys.path.insert(0, parent_dir)

import rospy
from sensor_msgs.msg import CompressedImage
from cv_bridge import CvBridge
import cv2 
import json
from std_msgs.msg import String
from drive_perception.camera.preprocessing import CameraPreprocessor
from drive_perception.camera.feature_extraction import LaneFeatureExtractor
from drive_perception.camera.sliding_window import SlidingWindow
from utills import check_timer

class PerCamera:
    # 차선 정보를 추출해 ROS 토픽으로 발행하는 메인 카메라 인지 노드
    def __init__(self):
        # ROS 노드 초기화
        rospy.init_node("per_camera_node")
        print(f"PerCamera start")
        self.init_pubSub()
        self.init_msg()
        self.init_processing()
        self.init_timer()

    def init_pubSub(self):
        # 카메라 센서에서 JPEG 압축 영상을 구독하고 차선 메시지를 발행
        rospy.Subscriber("/usb_cam/image_raw", CompressedImage, self.CB_cam_raw, queue_size=1)
        self.pub = rospy.Publisher('/perception/camera', String, queue_size=1)
    def init_msg(self):
        # OpenCV 이미지 객체 및 브리지 초기화
        self.img = None
        self.bridge = CvBridge()
    def init_processing(self):
        # 차선 검출 파이프라인에서 사용하는 전처리, 특징 추출, 창 탐색 객체 준비
        self.CameraPreprocessor = CameraPreprocessor()
        self.LaneFeatureExtractor = LaneFeatureExtractor()
        self.SlidingWindow = SlidingWindow()
    def init_timer(self):
        # 처리 주기 모니터링용 타이머
        self.check_timer = check_timer.CheckTimer("per_camera_node")

    def CB_cam_raw(self,msg):
        # 센서 이미지 콜백: 압축 이미지를 OpenCV 형식으로 변환 후 처리 파이프라인 실행
        self.img = self.bridge.compressed_imgmsg_to_cv2(msg)
        self.processing()

    def pub_cam_info(self,dataset):
        # 추출한 차선 정보를 JSON 문자열로 직렬화하여 발행
        json_str = json.dumps(dataset)
        self.pub.publish(json_str)
        
    def view_cam(self):
        # 디버깅용 시각화: 검출된 차선과 정지선 정보를 한 화면에 표시
        combined_img = cv2.bitwise_or(self.white_lane_img,self.yellow_lane_img)
        if self.stop_line != []:
            cross_threshold = 35
            min_y, max_y = self.stop_line
            cross_diff = (max_y - min_y)
            if cross_threshold < cross_diff:
                cv2.rectangle(combined_img,[0,min_y],[self.img_x, max_y],[0,0,255],3)
        #cv2.imshow("white_lane_img",white_lane_img)
        cv2.imshow("lane_img",combined_img)
        # cv2.imshow("self.img",self.img)
        cv2.waitKey(1)
        # end = time()
        # print(f"time1 {end - start1} ")
        
    def processing(self):
        try:
            # 1) 원본 이미지 크기를 구하고 BEV(Bird's Eye View) 변환으로 차선 평면화
            self.img_y, self.img_x = self.img.shape[0:2]
            warped_img = self.CameraPreprocessor.BEV_img_warp(self.img,self.img_y,self.img_x)
            # 2) HSV 공간에서 노란색/흰색 차선 후보를 추출할 수 있도록 색 필터링 준비
            warped_img_hsv = cv2.cvtColor(warped_img,cv2.COLOR_BGR2HSV)
            yellow_filtered_img, white_filtered_img = self.CameraPreprocessor.detect_color_yAndw(warped_img,warped_img_hsv)
            yellow_bin_img,white_bin_img = self.CameraPreprocessor.img_binary_yAndw(yellow_filtered_img, white_filtered_img)

            # 3) 슬라이딩 윈도우 파라미터 설정 및 차선 종류/정지선 추정
            self.SlidingWindow.set_img_y(self.img_y)
            self.lane_mode = self.LaneFeatureExtractor.estimate_lane_mode(warped_img)
            self.stop_line, white_bin_img= self.LaneFeatureExtractor.estimate_stop_line(white_bin_img,self.img_y)
            # 4) 좌우 차선을 각각 슬라이딩 윈도우로 추적해 궤적 점을 수집
            self.yellow_lane_img, yellow_left_lane, yellow_right_lane = self.SlidingWindow.sliding_window_adaptive(yellow_bin_img)
            self.white_lane_img, white_left_lane, white_right_lane = self.SlidingWindow.sliding_window_adaptive(white_bin_img)
            
            # 5) 정지선, 좌우 차선, 차선 모드 정보를 패키징 후 발행
            dataset = [self.stop_line, yellow_left_lane, yellow_right_lane,white_left_lane, white_right_lane,self.lane_mode]           
            self.pub_cam_info(dataset)
            
            # 6) 결과 시각화 (필요 시)
            self.view_cam()
        except Exception as e:
            pass
     
        
if __name__ == '__main__':
    node = PerCamera()
    rospy.spin()
    
# ---------------------------------------------------------------------------
# 전체 알고리즘 개요
# 1) 노드 초기화: ROS 노드/퍼블리셔/구독자를 구성하고 카메라 전처리, 특징 추출, 슬라이딩 윈도우 객체를 준비한다.
# 2) 이미지 수신: 콤프레스트 이미지를 수신하면 CvBridge로 OpenCV 이미지로 변환한 뒤 처리 파이프라인을 호출한다.
# 3) BEV 변환: 원근을 제거한 탑뷰(BEV) 이미지를 생성하여 차선 검출이 안정적으로 이루어지도록 한다.
# 4) 색상 분리: HSV 색공간으로 변환한 뒤 노란/흰 차선을 각각 강조하는 필터를 적용한다.
# 5) 이진화: 필터링된 결과를 바이너리 이미지로 변환하여 명확한 차선 픽셀 후보를 만든다.
# 6) 특징 추출: 차선 종류(차로수/패턴)와 정지선 위치를 추정하고, 이미지 높이를 슬라이딩 윈도우에 전달한다.
# 7) 슬라이딩 윈도우 추적: 노란색·흰색 이진 영상에서 좌우 차선을 각각 추적하며 궤적 좌표와 시각화 이미지를 만든다.
# 8) 메시지 발행: 정지선, 좌우 차선 궤적, 차선 모드 정보를 JSON 배열로 묶어 ROS 토픽으로 발행한다.
# 9) 시각화: 디버깅을 위해 추적된 차선과 정지선을 하나의 화면으로 띄워 실시간 검증이 가능하도록 한다.
# 예외 상황 발생 시 try/except로 전체 노드가 중단되지 않도록 안전하게 무시한다.

File: src/autorace_perception/src/lane_detection/src/per_camera.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/image_encodings.h>
#include <std_msgs/String.h>

#include <cv_bridge/cv_bridge.h>
#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include <cstdint>
#include <sstream>
#include <string>
#include <vector>

#include "lane_detection/feature_extraction.hpp"
#include "lane_detection/preprocessing.hpp"
#include "lane_detection/sliding_window.hpp"

namespace lane_detection {

namespace {

// 주기 측정/디버깅용 더미 클래스 (현재는 기능 비활성화 상태)
class CheckTimer {
public:
    explicit CheckTimer(const std::string& /*name*/) {}
    void tick(const std::string& /*phase*/ = std::string()) {}
};

std::string pointsToJson(const std::vector<cv::Point>& points) {
    std::ostringstream oss;
    oss << "[";
    for (size_t i = 0; i < points.size(); ++i) {
        oss << "[" << points[i].x << "," << points[i].y << "]";
        if (i + 1 < points.size()) {
            oss << ",";
        }
    }
    oss << "]";
    return oss.str();
}

std::string stopLineToJson(const std::vector<int>& stop_line) {
    if (stop_line.size() != 2) {
        return "[]";
    }
    std::ostringstream oss;
    oss << "[" << stop_line[0] << "," << stop_line[1] << "]";
    return oss.str();
}

cv::Mat toCvMat(const sensor_msgs::ImageConstPtr& msg) {
    if (!msg) {
        return cv::Mat();
    }
    try {
        auto cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
        return cv_ptr->image;
    } catch (const cv_bridge::Exception& e) {
        ROS_ERROR_STREAM_THROTTLE(1.0, "cv_bridge exception: " << e.what());
        return cv::Mat();
    }
}

}  // namespace

// 실시간 카메라 토픽을 구독해 BEV 변환 → 전처리 → 차선/정지선 검출 → 결과 퍼블리시까지 수행하는 ROS 노드
class PerCameraNode {
public:
    PerCameraNode()
        : nh_(),
          preprocessor_(),
          feature_extractor_(),
          sliding_window_(),
          check_timer_("per_camera_node"),
          img_x_(0),
          img_y_(0) {
        sub_ = nh_.subscribe("/usb_cam/image_rect_color", 1, &PerCameraNode::imageCallback, this);
        pub_ = nh_.advertise<std_msgs::String>("/perception/camera", 1);
        ROS_INFO("PerCamera C++ node started");
    }

    void spin() const {
        ros::spin();
    }

private:
    // USB 카메라 토픽으로부터 이미지가 들어올 때마다 호출
    void imageCallback(const sensor_msgs::ImageConstPtr& msg) {
        try {
            img_ = toCvMat(msg);
            if (img_.empty()) {
                ROS_WARN_THROTTLE(1.0, "Failed to convert image");
                return;
            }
            processing();
        } catch (const cv::Exception& e) {
            ROS_ERROR_STREAM("OpenCV exception: " << e.what());
        } catch (const std::exception& e) {
            ROS_ERROR_STREAM("Exception in image callback: " << e.what());
        }
    }

    // 단일 프레임에 대해 BEV 변환→색상 분리→이진화→정지선·차선 추출까지 수행
    void processing() {
        if (img_.empty()) {
            return;
        }

        try {
            img_y_ = img_.rows;
            img_x_ = img_.cols;

            cv::Mat warped_img = preprocessor_.bevWarp(img_, img_y_, img_x_);
            if (warped_img.empty()) {
                return;
            }

            cv::Mat warped_hsv;
            cv::cvtColor(warped_img, warped_hsv, cv::COLOR_BGR2HSV);

            cv::Mat yellow_filtered;
            cv::Mat white_filtered;
            preprocessor_.detectColorYellowAndWhite(warped_img,
                                                    warped_hsv,
                                                    yellow_filtered,
                                                    white_filtered);

            cv::Mat yellow_bin;
            cv::Mat white_bin;
            preprocessor_.binaryImagesYellowAndWhite(yellow_filtered,
                                                     white_filtered,
                                                     yellow_bin,
                                                     white_bin);

            sliding_window_.setImageHeight(img_y_);

            stop_line_ = feature_extractor_.estimateStopLine(white_bin, img_y_);
            lane_mode_ = feature_extractor_.estimateLaneMode(warped_img);

            auto yellow_result = sliding_window_.slidingWindowAdaptive(yellow_bin);
            auto white_result = sliding_window_.slidingWindowAdaptive(white_bin);

            yellow_lane_img_ = yellow_result.visualized;
            white_lane_img_ = white_result.visualized;
            yellow_left_lane_ = yellow_result.left_lane;
            yellow_right_lane_ = yellow_result.right_lane;
            white_left_lane_ = white_result.left_lane;
            white_right_lane_ = white_result.right_lane;

            publishCameraInfo();
            viewCamera();
        } catch (const cv::Exception& e) {
            ROS_ERROR_STREAM_THROTTLE(1.0, "Processing OpenCV exception: " << e.what());
        } catch (const std::exception& e) {
            ROS_ERROR_STREAM_THROTTLE(1.0, "Processing exception: " << e.what());
        }
    }

    void publishCameraInfo() {
        std::ostringstream oss;
        oss << "["
            << stopLineToJson(stop_line_) << ","
            << pointsToJson(yellow_left_lane_) << ","
            << pointsToJson(yellow_right_lane_) << ","
            << pointsToJson(white_left_lane_) << ","
            << pointsToJson(white_right_lane_) << ","
            << "\"" << lane_mode_ << "\""
            << "]";

        std_msgs::String msg;
        msg.data = oss.str();
        pub_.publish(msg);
    }

    void viewCamera() {
        if (yellow_lane_img_.empty() && white_lane_img_.empty()) {
            return;
        }

        cv::Mat combined;
        if (!yellow_lane_img_.empty() && !white_lane_img_.empty()) {
            cv::bitwise_or(yellow_lane_img_, white_lane_img_, combined);
        } else if (!yellow_lane_img_.empty()) {
            combined = yellow_lane_img_.clone();
        } else {
            combined = white_lane_img_.clone();
        }

        if (stop_line_.size() == 2) {
            int min_y = stop_line_[0];
            int max_y = stop_line_[1];
            int cross_diff = max_y - min_y;
            if (cross_diff > 35) {
                cv::rectangle(combined,
                              cv::Point(0, min_y),
                              cv::Point(std::max(0, img_x_ - 1), max_y),
                              cv::Scalar(0, 0, 255),
                              3);
            }
        }
        
        cv::imshow("lane_img", combined);
        cv::waitKey(1);
    }

    ros::NodeHandle nh_;
    ros::Subscriber sub_;
    ros::Publisher pub_;

    CameraPreprocessor preprocessor_;
    LaneFeatureExtractor feature_extractor_;
    SlidingWindow sliding_window_;
    CheckTimer check_timer_;

    cv::Mat img_;
    cv::Mat yellow_lane_img_;
    cv::Mat white_lane_img_;

    std::vector<int> stop_line_;
    std::vector<cv::Point> yellow_left_lane_;
    std::vector<cv::Point> yellow_right_lane_;
    std::vector<cv::Point> white_left_lane_;
    std::vector<cv::Point> white_right_lane_;
    std::string lane_mode_;

    int img_x_;
    int img_y_;
};

}  // namespace lane_detection

int main(int argc, char** argv) {
    ros::init(argc, argv, "per_camera_node");  // 노드 초기화
    lane_detection::PerCameraNode node;        // 주요 처리 클래스 생성
    node.spin();                               // ROS 이벤트 루프 실행
    return 0;
}